elements/rh-pagination/rh-pagination.ts

```diff
@@ -0,0 +1,179 @@
+
+  #screenSize = new ScreenSizeController(this);
+  #logger = new Logger(this);
+
+  #nav = this.querySelector('nav');
+  #links = this.#nav?.querySelectorAll<HTMLAnchorElement>('li a');
+  #currentLink = this.#getCurrentLink();
+
+  #currentIndex = -1;
+  #firstLink: HTMLAnchorElement | null = null;
+  #lastLink: HTMLAnchorElement | null = null;
+  #nextLink: HTMLAnchorElement | null = null;
+  #prevLink: HTMLAnchorElement | null = null;
+
+  update(changed: PropertyValues<this>): void {
+    this.#update();
+    super.update(changed);
+  }
+
+  render() {
+    const { mobile, size } = this.#screenSize;
+    const { labelFirst, labelPrevious, labelNext, labelLast } = this;
+    const firstHref = this.#currentLink === this.#firstLink ? undefined : this.#firstLink?.href;
+    const prevHref = this.#prevLink?.href;
+    const nextHref = this.#nextLink?.href;
+    const lastHref = this.#currentLink === this.#lastLink ? undefined : this.#lastLink?.href;
+    return html`
+      <div id="container" class=${classMap({ mobile, [size as string]: true })}>
+        <a id="first" class="stepper" href=${ifDefined(firstHref)} aria-label=${labelFirst}>${L2}</a>
+        <a id="prev" class="stepper" href=${ifDefined(prevHref)} aria-label=${labelPrevious}>${L1}</a>
+
+        <div id="nav-container" ?hidden=${mobile}>
+          <slot @slotchange=${this.#update}></slot>
+        </div>
+
+        <a id="next" class="stepper" href=${ifDefined(nextHref)} aria-label=${labelNext}>${L1}</a>
+        <a id="last" class="stepper" href=${ifDefined(lastHref)} aria-label=${labelLast}>${L2}</a>
+
+        <div id="numeric" ?hidden=${!mobile} ?inert=${!mobile}>
+          <span id="go-to-page">
+            <slot name="go-to-page">Go to page</slot>
+          </span>
+          <input inputmode="numeric" aria-labelledby="go-to-page" value=${(this.#currentIndex ?? 0) + 1} />
+          <slot name="out-of">of</slot>
+          <a href=${ifDefined(lastHref)}>${this.#links?.length}</a>
+        </div>
+      </div>
+    `;
+  }
+
+  #update() {
+    this.#updateLightDOMRefs();
+    this.overflow = this.#getOverflow();
+    this.#validateA11y();
+  }
+
+  #getOverflow(): 'start' | 'end' | 'both' | null {
+    const overflowAt = 9;
+    const length = this.#links?.length ?? 0;
+    if (length <= overflowAt) {
+      return null;
+    }
+
+    const current = this.#currentIndex + 1;
+
+    if (current > (overflowAt - 4) && current < (length - 4)) {
+      return 'both';
+    } else if (current <= (overflowAt - 4)) {
+      return 'end';
+    } else {
+      return 'start';
+    }
+  }
+
+  #getCurrentLink(): HTMLAnchorElement | null {
+    const ariaCurrent = this.querySelector<HTMLAnchorElement>('li a[aria-current="page"]');
+    if (ariaCurrent) {
+      return ariaCurrent;
+    }
+    for (const link of this.#links ?? []) {
+      const url = new URL(link.href);
+      if (url.pathname === location.pathname && url.search === location.search && url.hash === location.hash) {
```

Comment:
These might not be valid urls if they are in an SPA.  Like if they do the "#"'s.

Reply:
I added some more facility for SPAs. check the readme and LMK what you think. The demo is basically an SPA. I also added some buttons to the demo which add/remove pages (like your MO case from above)

Reply:
ahh, I'm sorry.  I missunderstood what this was doing.  I thought it was also validating the URL.  The docs explain perfectly. 👍